text/0000-transaction-refactoring.md

```diff
@@ -0,0 +1,327 @@
+
+The new `Receipt` contains the shared information and either one of the receipt actions or a list of actions:
+- `sender_id` the account ID of the immediate previous sender of this receipt. It can be different from the `originator_id` in some cases.
+- `receiver_id` the account ID of the current account, on which we need to perform action(s).
+- `receipt_id` is a hash of this receipt (previously was called `nonce`). It's generated from either the signed transaction or the parent receipt.
+- `receipt` is can be one of 2 types:
+  - `ActionReceipt` is used to perform some actions on the receiver.
+  - `DataReceipt` is used when some data needs to be passed from the sender to the receiver, e.g. an execution result.
+
+To support promises and callbacks we introduce a concept of cross-shard data sharing with dependencies. Each `ActionReceipt` may have a list of input `data_id`. The execution will not start until all required inputs are received. Once the execution completes and if there is `output_data_id`, it produces a `DataReceipt` that will be routed to the `output_receiver_id`.
+ 
+`ActionReceipt` contains the following fields:
+- `originator_id` the account ID of the originator, who signed the transaction.
+- `originator_public_key` the public key that the originator used to sign the original signed transaction.
+- `refund_account_id` the account ID where to send the refund in case the transaction fails and/or there are remaining fees.
+- `output_data_id` is the data ID to create DataReceipt. If it's absent, then the `DataReceipt` is not created.
+- `output_receiver_id` is the account ID of the data receiver. It's needed to route `DataReceipt`. It's absent if the DataReceipt is not needed.
+- `input_data_id` is the list of data IDs that are required for the execution of the `ActionReceipt`. If some of data IDs is not available when the receipt is received, then the `ActionReceipt` is postponed until all data is available. Once the last `DataReceipt` for the required input data arrives, the action receipt execution is triggered.
+- `action` is the list of actions to execute. The execution doesn't need to validate permissions of the actions, but need to fail in some cases. E.g. when the receiver's account doesn't exist and the action acts on the account, or when the action is a function call and the code is not present.
+
+`DataReceipt` contains the following fields:
+- `data_id` is the data ID to be used as an input.
+- `success` is true if the `ActionReceipt` that generated this `DataReceipt` finished the execution without any failures.
+- `data` is the binary data that is returned from the last action of the `ActionReceipt`. Right now, it's empty for all actions except for function calls. For function calls the data is the result of the code execution. But in the future we might introduce non-contract state reads.
+
+Data should be stored at the same shard as the receiver's account, even if the receiver's account doesn't exist.
```

Comment:
What's a scenario where this would happen?

Reply:
For callbacks on other shards, will add an example